services/QuillLMS/app/models/session_feedback_history.rb

```diff
@@ -0,0 +1,87 @@
+      .joins("LEFT OUTER JOIN comprehension_prompts ON feedback_histories.prompt_id = comprehension_prompts.id")
+      .joins("LEFT OUTER JOIN feedback_histories AS because_feedback ON feedback_histories.id = because_feedback.id AND comprehension_prompts.conjunction = 'because'")
+      .joins("LEFT OUTER JOIN feedback_histories AS but_feedback ON feedback_histories.id = but_feedback.id AND comprehension_prompts.conjunction = 'but'")
+      .joins("LEFT OUTER JOIN feedback_histories AS so_feedback ON feedback_histories.id = so_feedback.id AND comprehension_prompts.conjunction = 'so'")
+      .where(used: true)
+      .group(:activity_session_uid, :activity_id)
+      .order('start_date DESC')
+    query = query.where(comprehension_prompts: {activity_id: activity_id.to_i}) if activity_id
+    query = query.limit(page_size)
+    query = query.offset((page - 1) * page_size.to_i) if page && page > 1
+    query
+  end
+
+  def self.serialize_list_by_activity_session(results)
+    results.map do |r|
+      {
+        session_uid: r.session_uid,
+        start_date: r.start_date.iso8601,
+        activity_id: r.activity_id,
+        because_attempts: r.because_attempts || 0,
+        but_attempts: r.but_attempts || 0,
+        so_attempts: r.so_attempts || 0,
+        complete: r.complete || false
+      }
+    end
+  end
+
+  def self.serialize_detail_by_activity_session(activity_session_uid)
+    histories = FeedbackHistory.where(activity_session_uid: activity_session_uid).all
+
+    return nil if histories.empty?
+
+    start_date = histories.first&.time
+    activity_id = histories.first&.prompt&.activity_id
+    because_attempts, because_complete = serialize_conjunction_feedback_history(histories, 'because')
+    but_attempts, but_complete = serialize_conjunction_feedback_history(histories, 'but')
+    so_attempts, so_complete = serialize_conjunction_feedback_history(histories, 'so')
+
+    output = {
+      start_date: start_date&.iso8601,
+      session_uid: activity_session_uid,
+      activity_id: activity_id,
+      session_completed: because_complete && but_complete && so_complete,
+      prompts: []
+    }
+
+    output[:prompts].push(because_attempts) if because_attempts[:prompt_id]
+    output[:prompts].push(but_attempts) if but_attempts[:prompt_id]
+    output[:prompts].push(so_attempts) if so_attempts[:prompt_id]
+
+    output
+  end
+
+  private_class_method def self.serialize_conjunction_feedback_history(feedback_histories, conjunction=nil)
+    feedback_histories = feedback_histories.filter { |h| h.prompt&.conjunction == conjunction } if conjunction
+
+    conjunction_prompt = {
+      prompt_id: feedback_histories.first&.prompt_id,
+      conjunction: feedback_histories.first&.prompt&.conjunction,
+      attempts: {}
+    }
+
+    feedback_histories.each do |feedback_history|
```

Comment:
It seems like you are bouncing between an array and a hash for `attempts`. Couldn't you just select the attempts for a prompt and sort by attempt number? That would give you the array.

Reply:
We could do that, but because we include values where "used" is `false`, we'll end up with multiple entries for each attempt number.  That probably doesn't matter for our purposes (I don't think the ambiguity is problematic in any of our use-cases), so this may be worth shifting toward.  The current spec calls for a hash of arrays, but we could just make it a single array instead.